Android improvements

kormang opened the pull request:
Simplified android code responsible for adding and removing children to ReactNativeMapboxGLView. This should also improve performance, although there was no performance measuring.

Also there was a bug in getBounds method, which is now fixed in relatively acceptable way.

kormang committed:
Refactored android child and custom annotation handling

kormang committed:
Fixed bug on android when bounds are requested and _map is null

If bounds are requested and onMapReady is not yet called, _map is null.
Previously in such situation LatLngBounds are creater using builder but without
any LatLng item, so builder would throw an exception saying "Can not create LatLngBounds with 0 items". Now initial center LatLng is repeated twice, since two items are required. This is still better then throwing exception that user can not understand or control.

henrikra commented on android/src/main/java/com/mapbox/reactnativemapboxgl/ReactNativeMapboxGLView.java:
```diff
@@ -682,7 +726,13 @@ public CameraPosition getCameraPosition() {
+            LatLng center = _initialCamera.build().target;
+            return new LatLngBounds.Builder()
+                    .include(center)
+                    .include(center)
```
Why `.include(center)` is twice here?

kormang replied:
Oh, it's because including center twice works, there is no problem with that. We need two points provided to builder, and it is not complaining if both are same. I can leave comment in code if it will make difference.

kormang replied:
Without it this code will throw exception. I though this is acceptable way to fix it. At least it works for me, and it is certainly better that getting red screen with stack trace printed and having no way of controlling it.

henrikra replied:
Yeah I think it should be commented then because it looks like a mistake. There is no other way of doing it?

kormang replied:
Maybe there is, but I think it requires broader discussion about behavior of the whole system. Maybe we should prevent this method of even being called when _map is null. But as I said it might require broader discussion about it, and larger refactoring. As I'm using MapboxGL, I'm open for cooperation, and would like to make better pull request. Again, this is certainly better that it used to be, and in my project it works in acceptable way.

kormang committed:
A comment explaining getBounds in ReactNativeMapboxGLView.java when _map is null

kormang committed:
Checking if _mapView is null in onHostResume as it sometimes can be

kormang committed:
Merge branch 'master' of github.com:mapbox/react-native-mapbox-gl into android_improvements

kormang committed:
Emmiting ON_FINISH_LOADING_MAP event when map finishes loading AND map is ready

Solves the issue #586. When onMapFinishedLoading is called user would request LatLng bounds,
but it is not always possible at that points since onMapReady in java code is still not called,
and _map is null. Previouse in that case user would get ugly red screen with exception displayed,
and could do nothing about it. Now user can be sure that when onMapFinishedLoading is called,
map is completely ready.

I don't like introducing new flags, but could not think about better solution.
At least two flags introduced in my commits are self describing, and their purpose is clear.

kormang committed:
Fixed sometimes missing custom annotations

kormang committed:
Merge branch 'master' into android_improvements
